# Analyze the Risk of the Borrower Failing to Pay

### Table of Contents

* [Analyze the Risk of the Borrower Failing to Pay](#scrollTo=m2KyziSkg3Ta)
    * [Pre-processing Data](#scrollTo=AHuel4Iaf9OR)
    * [1. Open file Data and Display General Information.](#scrollTo=BJ_98FqNf9OX)
    * [2. Data Preprocessing](#scrollTo=1Tdf9ve4f9Ob)
        * [2.1. Identifying Missing Values](#scrollTo=CL3ERkhywZf0)
        * [2.2. Data Transformation](#scrollTo=-JRdn7BTf9Oq)
            * [2.2.1. Checking the education Column](#scrollTo=0UDtw2NMf9Or)
            * [2.2.2. Checking the children Column](#scrollTo=O8-NOwI_f9Ou)
            * [2.2.3. Checking the days_employed Column](#scrollTo=fyf8ElSEf9O4)
            * [2.2.4. Checking the dob_years Column](#scrollTo=FNDxGPjdf9O8)
            * [2.2.5. Checking the family_status Column](#scrollTo=qMByFGqZf9O_)
            * [2.3.6. Checking the gender Column](#scrollTo=6zpE6_uvf9PB)
            * [2.3.7. Checking the income_type Column](#scrollTo=yxLG9B-Yf9PE)
        * [2.3. Checking Duplicates](#scrollTo=xKEid7pjyLRl)
        * [2.4. Work with Missing Values](#scrollTo=h8DWab5Jf9PM)
            * [2.4.1. Fixed Missing Values in total_income](#scrollTo=kmfC7wCtf9PO)
            * [2.4.2. Fixed the Missing Values in days_employed](#scrollTo=1G6BLS4cf9PX)
        * [2.5. Data Categorization](#scrollTo=R5-x_yR0f9Pb)
            * [2.5.1. Categorizing Data in the purpose Column](#scrollTo=Q7alM5MJy0Ae)
            * [2.5.2. Categorizing Data in the total_income Column](#scrollTo=kTVT22SPzOWO)
    * [3. Answering the Hypothesis](#scrollTo=Z-a1SKx5f9Ph)
        * [3.1. Is there a relationship between having children and the probability of defaulting on loans?](#scrollTo=8Obi_s-Qf9Ph)
        * [3.2. Is there a correlation between family status and paying back on time?](#scrollTo=P3sJMUaBf9Pk)
        * [3.3. Is there a correlation between income levels and paying back on time?](#scrollTo=0dru-vduf9Pm)
        * [3.4. How do different loan objectives affect the probability of defaulting on a loan??](#scrollTo=bdoMQRgof9Po)
    * [4. General Conclusion](#scrollTo=AkaFtVnSf9Pq)



## Pre-processing Data

As a credit analyst, our project is to prepare reports for the bank's credit department. We find out the effect of a customer's marital status and the number of children on the probability of being on time in repaying the loan. Banks already have some data regarding customer creditworthiness.

Our report will be considered when making a **credit assessment** for prospective customers. **Credit scoring** is used to evaluate the ability of potential borrowers to repay their loans.

The main purpose of this project is to determine a client's eligibility to get credit based on their status and circumstances stored in our data. We also test the capacity of customers based on their characteristics which we summarize based on categories so that *patterns* are obtained to give yellow lights to customers who fall into certain categories.

Project hypothesis:
1. Is there a correlation between the number of children and the ability to repay loans on time?
2. Is there a correlation between family status and the ability to repay loans on time?
3. Is there a correlation between economic class and the ability to repay loans on time?
4. Is there a correlation between credit goals and the ability to repay loans on time?

## 1. Open *file* Data and Display General Information.

We'll start by importing the library and loading the data.

In [ ]:
# Load all libraries
import pandas as pd

In [ ]:
# load data
df = pd.read_csv('https://practicum-content.s3.us-west-1.amazonaws.com/datasets/credit_scoring_eng.csv')

## 2. Data Preprocessing

**Data Description**
- *`children`* - the number of children in the family
- *`days_employed`* - work experience in days
- *`dob_years`* - client's age in years
- *`education`* - client education
- *`education_id`* - educational ID
- *`family_status`* - marital status
- *`family_status_id`* - marital status identifier
- *`gender`* - the gender of the client
- *`income_type`* - type of work
- *`debt`* - does the client have loan payment debts
- *`total_income`* - monthly income
- *`purpose`* - the purpose of getting a loan

In [ ]:
# Check the number of rows and columns in the dataset
df.shape

(21525, 12)

In [ ]:
# Display the first 10 rows in the dataset
df.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

From the sample data displayed, several problems can be detected. Namely, there are negative values, and the value we value is very high from the `days_employed` column, which we value is unreasonable because the column displays work experience in days, as well as writing capital letters that are not incorrect regularly in the `education` column.

In [ ]:
# Get the information of all columns in the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


There are missing values in the `days_employed` and `total_income` columns.

In [ ]:
# Display missing values in the dataset
df[df['days_employed'].isna()].head(10)

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   
65         0            NaN         21  secondary education             1   
67         0            NaN         52    bachelor's degree             0   
72         1            NaN         32    bachelor's degree             0   
82         2            NaN         50    bachelor's degree             0   
83         0            NaN         52  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   
65          unmarried                 4      M       business     0   
67            married                 0      F        retiree     0   
72            married                 0      M  civil servant     0   
82            married                 0      F       employee     0   
83            married                 0      M       employee     0   

    total_income                                   purpose  
12           NaN                         to have a wedding  
26           NaN                                 education  
29           NaN                    building a real estate  
41           NaN                  second-hand car purchase  
55           NaN                         to have a wedding  
65           NaN  transactions with commercial real estate  
67           NaN       purchase of the house for my family  
72           NaN  transactions with commercial real estate  
82           NaN                                   housing  
83           NaN                                   housing

### 2.1. Identifying Missing Values

So far, from what I've seen, the missing values appear symmetrical because they are on the same line. Still, to conclude whether the missing values are on the same line, further investigation is needed.

In [ ]:
# Perform some filtering to find out the number of rows of missing values
df_filtered_nan = df[df['days_employed'].isna()]
df_filtered_nan = df_filtered_nan[df_filtered_nan['total_income'].isna()]
df_filtered_nan.shape[0]

2174

The missing data in our database is `2174` rows.

In [ ]:
# Now we will calculate the missing value ratio of the whole dataframe
df_distribution_nan = df_filtered_nan.shape[0] / df.shape[0] 
print(f'Distribusi nilai yang hilang sebesar: {df_distribution_nan:0%}')

Distribusi nilai yang hilang sebesar: 10.099884%


**Intermediate conclusion**

The number of missing values is equal to the number of tables filtered. It means the missing values from the filtered table are symmetrical.

The percentage of missing data from the dataframe is `10%`, affecting our data.

Next, we will calculate the percentage of missing values, and whether they significantly impact the data in the dataset so we will know the right steps to process the missing data.

In [ ]:
# Apply a filter to display missing rows from the data
df_filtered_nan.head(10)

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   
65         0            NaN         21  secondary education             1   
67         0            NaN         52    bachelor's degree             0   
72         1            NaN         32    bachelor's degree             0   
82         2            NaN         50    bachelor's degree             0   
83         0            NaN         52  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   
65          unmarried                 4      M       business     0   
67            married                 0      F        retiree     0   
72            married                 0      M  civil servant     0   
82            married                 0      F       employee     0   
83            married                 0      M       employee     0   

    total_income                                   purpose  
12           NaN                         to have a wedding  
26           NaN                                 education  
29           NaN                    building a real estate  
41           NaN                  second-hand car purchase  
55           NaN                         to have a wedding  
65           NaN  transactions with commercial real estate  
67           NaN       purchase of the house for my family  
72           NaN  transactions with commercial real estate  
82           NaN                                   housing  
83           NaN                                   housing

Missing values appear symmetrical from the displayed table.

In [ ]:
# Checking distribution
print(df_filtered_nan['income_type'].value_counts())
print()
df_filtered_nan['income_type'].value_counts() / df['income_type'].value_counts() * 100

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64



business                        9.990167
civil servant                  10.075394
employee                        9.937944
entrepreneur                   50.000000
paternity / maternity leave          NaN
retiree                        10.710581
student                              NaN
unemployed                           NaN
Name: income_type, dtype: float64

There is quite a pattern here, but there are indeed several categories whose values cannot be said to be valid for calculations considering that there are only a few '1 - 2' data, such as the 'unemployed, paternity/maternity leave, student,' and 'entrepreneur' categories. Overall the missing data is distributed by `10%` in each category.

**Possible causes of missing values in data**

The cause of the lost value cannot be determined yet. We must consider other possibilities for the cause of the lost value, whether the value has certain characteristics such as from married clients, number of children, or clients who have outstanding credit payments.

In [ ]:
# Checking the distribution across datasets
df.isna().sum() / df.shape[0] * 100

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64

**Tentative conclusions**

The number of missing values in a dataset is similar to a filtered table.
And the amount of missing data in the `days_employed` column is the same as in `total_income`, which means that the error only occurs in those two columns.

This raises the question of whether the missing values form a pattern or occur randomly?
To answer this question, let's do further analysis.

In [ ]:
# Check if there are other patterns that cause missing data from the 'family_status' column
print(df_filtered_nan['family_status'].value_counts())
print()
df_filtered_nan['family_status'].value_counts() / df['family_status'].value_counts() * 100

married              1237
civil partnership     442
unmarried             288
divorced              112
widow / widower        95
Name: family_status, dtype: int64



married               9.991922
civil partnership    10.581757
unmarried            10.238180
divorced              9.372385
widow / widower       9.895833
Name: family_status, dtype: float64

**Intermediate conclusion**

Interestingly, the missing values are evenly distributed across the categories in the `family_status` column, around `10%`. But it is also known that no special categories result in lost data caused by just one category.

In [ ]:
# Search for relationship 'gender' and missing value
print(df_filtered_nan['gender'].value_counts())
print()
df_filtered_nan['gender'].value_counts() / df['gender'].value_counts() * 100

F    1484
M     690
Name: gender, dtype: int64



F      10.424276
M       9.467618
XNA          NaN
Name: gender, dtype: float64

In the gender category, the missing data is distributed respectively by `9%` and `10%`, meaning that the missing data is not only found in one category.

In [ ]:
# Checking income from some jobs that may not have 'income'
df[df['income_type'].isin(['unemployed', 'paternity / maternity leave', 'student', 'entrepreneur'])]

children  days_employed  dob_years            education  education_id  \
3133          1  337524.466835         31  secondary education             1   
5936          0            NaN         58    bachelor's degree             0   
9410          0    -578.751554         22    bachelor's degree             0   
14798         0  395302.838654         45    Bachelor's Degree             0   
18697         0    -520.848083         27    bachelor's degree             0   
20845         2   -3296.759962         39  SECONDARY EDUCATION             1   

           family_status  family_status_id gender  \
3133             married                 0      M   
5936             married                 0      M   
9410           unmarried                 4      M   
14798  civil partnership                 1      F   
18697  civil partnership                 1      F   
20845            married                 0      F   

                       income_type  debt  total_income  \
3133                    unemployed     1      9593.119   
5936                  entrepreneur     0           NaN   
9410                       student     0     15712.260   
14798                   unemployed     0     32435.602   
18697                 entrepreneur     0     79866.103   
20845  paternity / maternity leave     1      8612.661   

                               purpose  
3133   buying property for renting out  
5936       buy residential real estate  
9410      construction of own property  
14798               housing renovation  
18697                 having a wedding  
20845                              car

From the table above, we can get information that lost value is only sometimes caused by jobs that usually don't have income. There are several reasons they get income even though they don't have a regular job. As in line `3133`, a client who is `unemployed` but applies for a loan to buy a property for rent, although it is not yet known where the income comes from, Client `9410` is a `student` whether he is a part-time worker or has *rich parents*. Still, it is quite strange considering its use is not to continue education but to build property. Apart from that, several countries also have regulations that `paternity/maternity leave` still getting paid.

**Conclusion**

The conclusion we can draw regarding the causes of missing data is *human error* because missing data occurs in certain categories and several categories.

From some of the tests I have done, I found that from the `family_status` and `income_type` columns I found that each category in each of these columns has almost the same missing value in each category which is around `10%` meaning the missing values are distributed evenly in each category, and the values are symmetrical with the test we did earlier by finding the distribution of missing values across the dataset which results in a number also equal to `10%`.

For some problems like:
1. For the missing values, I will create categories based on age to find the average `days_employed` and `total_income`, which will be used to fill in the missing values.
2. To address the different registers in the `education` column, I will change all letters to `lower`.
3. We can do a `drop` for duplicate values.

### 2.2. Data Transformation

#### 2.2.1. Checking the `education` Column

In [ ]:
# Checking the spelling in the 'education' column which indicates that there are differences in writing with the same meaning
df_education_value = df.pivot_table(index='education', values='days_employed', aggfunc= 'count')
df_education_value

days_employed
education                         
BACHELOR'S DEGREE              251
Bachelor's Degree              243
GRADUATE DEGREE                  1
Graduate Degree                  1
PRIMARY EDUCATION               16
Primary Education               14
SECONDARY EDUCATION            705
SOME COLLEGE                    22
Secondary Education            646
Some College                    40
bachelor's degree             4222
graduate degree                  4
primary education              231
secondary education          12342
some college                   613

In [ ]:
# Then we will check for duplicate values before we deal with irregular registers
df.duplicated().sum()

54

In [ ]:
# Improve writing
df['education'] = df['education'].str.lower()

In [ ]:
# Rechecking the duplicate value after repair
df.duplicated().sum()

71

The duplicate values increase from `54` rows to `71` rows, meaning that there are rows with the same value in the data with different `education` inputs.

In [ ]:
# Check if the writing of the 'education' column has been corrected
df_education_value = df.pivot_table(index='education', values='days_employed', aggfunc= 'count')
df_education_value

days_employed
education                         
bachelor's degree             4716
graduate degree                  6
primary education              261
secondary education          13693
some college                   675

We have fixed the problem with the `education` column. Next, we will examine the `children` column.

#### 2.2.2. Checking the `children` Column

In [ ]:
# Display the distribution of values in the `children` column
df_children_value = df.pivot_table(index='children', values='days_employed', aggfunc= 'count')
df_children_value

days_employed
children               
-1                   44
 0                12710
 1                 4343
 2                 1851
 3                  294
 4                   34
 5                    8
 20                  67

- Some values show the numbers -1 and 20, which are unlikely to occur in the age group.
- We assume the problematic value occurs due to an input/typo error.
- We will replace the value -1 to 1 and 20 to 2.

In [ ]:
# Checking the number and percentage of problematic data in the 'children' column
df_child_problem_value = (df['children'] == -1).sum() + (df['children'] == 20).sum()
print('Total unfair value:', df_child_problem_value)
df_child_prob_percent = df_child_problem_value / df.shape[0]
print(f'The percentage of problematic data is: {df_child_prob_percent:.2%}')

Total unfair value: 123
The percentage of problematic data is: 0.57%


Data with unreasonable values are not up to `1%` in our dataset, even so, we better fix it instead of *dropping* on that data.

In [ ]:
# Fixing problematic values
df.loc[df['children'] == -1, 'children'] = 1 
df.loc[df['children'] == 20, 'children'] = 2 
df.duplicated().sum()

71

In [ ]:
# Double check the `children` field to make sure everything is fixed
df_children_value = df.pivot_table(index='children', values='days_employed', aggfunc= 'count')
df_children_value

days_employed
children               
0                 12710
1                  4387
2                  1918
3                   294
4                    34
5                     8

In [ ]:
# Let's check the duplicate value again
df.duplicated().sum()

71

No changes to duplicate values yet.

#### 2.2.3. Checking the `days_employed` Column

In [ ]:
# We will see an unusual value in the 'days_employed' column
df[df['income_type'] == 'retiree']

children  days_employed  dob_years            education  education_id  \
4             0  340266.072047         53  secondary education             1   
12            0            NaN         65  secondary education             1   
18            0  400281.136913         53  secondary education             1   
24            1  338551.952911         57  secondary education             1   
25            0  363548.489348         67  secondary education             1   
...         ...            ...        ...                  ...           ...   
21505         0  338904.866406         53  secondary education             1   
21508         0  386497.714078         62  secondary education             1   
21509         0  362161.054124         59    bachelor's degree             0   
21518         0  373995.710838         59  secondary education             1   
21521         0  343937.404131         67  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
4      civil partnership                 1      F     retiree     0   
12     civil partnership                 1      M     retiree     0   
18       widow / widower                 2      F     retiree     0   
24             unmarried                 4      F     retiree     0   
25               married                 0      M     retiree     0   
...                  ...               ...    ...         ...   ...   
21505  civil partnership                 1      M     retiree     0   
21508            married                 0      M     retiree     0   
21509            married                 0      M     retiree     0   
21518            married                 0      F     retiree     0   
21521            married                 0      F     retiree     0   

       total_income                                   purpose  
4         25378.572                         to have a wedding  
12              NaN                         to have a wedding  
18         9091.804                  buying a second-hand car  
24        46487.558  transactions with commercial real estate  
25         8818.041                           buy real estate  
...             ...                                       ...  
21505     12070.399                         to have a wedding  
21508     11622.175                                  property  
21509     11684.650                  real estate transactions  
21518     24618.344                         purchase of a car  
21521     24959.969                         purchase of a car  

[3856 rows x 12 columns]

In [ ]:
# Display descriptive statistics on the 'days_employed' column
df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

There are negative numbers and unreasonable values; please note this column displays the number of days the client has worked.

In [ ]:
# Display the average value of 'days_employed'
df_days_employed_median = df.groupby(['income_type'])['days_employed'].median()
df_days_employed_median

income_type
business                        -1547.382223
civil servant                   -2689.368353
employee                        -1574.202821
entrepreneur                     -520.848083
paternity / maternity leave     -3296.759962
retiree                        365213.306266
student                          -578.751554
unemployed                     366413.652744
Name: days_employed, dtype: float64

We cannot carry out the creditworthiness test process before correcting the value in this column. A negative value is the same as a missing value.

In [ ]:
# displays the ratio of problematic data
df_days_employed_problem_value = (df['days_employed'] < 0).sum() + (df['days_employed'] > 21600).sum()
print(df_days_employed_problem_value)
df_days_employed_prob_percent = df_days_employed_problem_value / df.shape[0]
print(f'The percentage of problematic data is: {df_days_employed_prob_percent:.2%}')

19351
The percentage of problematic data is: 89.90%


- There is a very high problem in the `days_employed` column; if we see a negative value, it has a big impact on the dataset because it is the same as a missing value. Then what about a very high value? No one works beyond their own age.
- Since the values might occur due to a technical error in adding the `-` sign, I can only replace negative numbers with positive ones with the `absolute` function.
- Then, for values that are too high, which we assume only occur in `categories`, `retiree` and `unemployed`, we decide to replace these values with `0` with a limit of `21600 days or 59 years`.

In [ ]:
# Apply the 'absolute' function to negative values and 'replace' unnatural values with '0'
df['days_employed'] = df['days_employed'].abs()
df.loc[df['days_employed'] > 21600, 'days_employed'] = 0

In [ ]:
# Ensure grades have been corrected
print(df['days_employed'].describe())
print()
print((df['days_employed'] < 0).sum())

count    19351.000000
mean      1934.115623
std       2274.751213
min          0.000000
25%        291.095954
50%       1203.369529
75%       2747.423625
max      18388.949901
Name: days_employed, dtype: float64

0


There is a problem with the `age` column i.e. no one has worked since age `0`

#### 2.2.4. Checking the `dob_years` Column

In [ ]:
# Checking the 'dob_years' column for anomalous values
df_dob_years_value = df.pivot_table(index='dob_years', values='days_employed', aggfunc= 'count')
df_dob_years_value

days_employed
dob_years               
0                     91
19                    13
20                    46
21                    93
22                   166
23                   218
24                   243
25                   334
26                   373
27                   457
28                   446
29                   495
30                   482
31                   495
32                   473
33                   530
34                   534
35                   553
36                   492
37                   484
38                   544
39                   522
40                   543
41                   548
42                   532
43                   463
44                   503
45                   447
46                   427
47                   421
48                   492
49                   458
50                   463
51                   398
52                   431
53                   415
54                   424
55                   395
56                   433
57                   404
58                   405
59                   410
60                   338
61                   317
62                   314
63                   240
64                   228
65                   174
66                   163
67                   151
68                    90
69                    80
70                    62
71                    53
72                    31
73                     7
74                     6
75                     1

In [ ]:
# View the percentage of anomaly data
df_days_employed_prob_percent = (df['dob_years'] == 0).sum() / df.shape[0]
print(f'The percentage of problematic data is: {df_days_employed_prob_percent:.2%}')

The percentage of problematic data is: 0.47%


We will replace the value `0` in the `age` column with the `median` age.
Because each client with an age of `0` has different characteristics.

In [ ]:
# Replace '0' with average
avg_dob_years = df['dob_years'].median() 
df.loc[df['dob_years'] == 0, 'dob_years'] = avg_dob_years

In [ ]:
# Memastikai nilai telah diperbaiki
df[df['dob_years'] == 0].shape[0]

0

#### 2.2.5. Checking the `family_status` Column

In [ ]:
# Checking the 'family_status' column
df_family_status_value = df.pivot_table(index='family_status', values='days_employed', aggfunc= 'count')
df_family_status_value

days_employed
family_status                   
civil partnership           3735
divorced                    1083
married                    11143
unmarried                   2525
widow / widower              865

No problems were found in the `family_status` column

#### 2.3.6. Checking the `gender` Column

In [ ]:
# Check "Gender" column
df_gender_value = df.pivot_table(index='gender', values='days_employed', aggfunc= 'count')
df_gender_value

days_employed
gender               
F               12752
M                6598
XNA                 1

There is one problem with the value `XNA`
It's difficult to identify where to replace the value, and we decided to delete that 1 row

In [ ]:
# Removing the value `XNA`
df = df.loc[df["gender"] != 'XNA']

In [ ]:
# Memastikan kolom telah diperbaiki
df_gender_value = df.pivot_table(index='gender', values='days_employed', aggfunc= 'count')
df_gender_value

days_employed
gender               
F               12752
M                6598

#### 2.3.7. Checking the `income_type` Column

In [ ]:
# Let's look at the values in the column
df_income_type_value = df.pivot_table(index='income_type', values='days_employed', aggfunc= 'count')
df_income_type_value

days_employed
income_type                               
business                              4576
civil servant                         1312
employee                             10014
entrepreneur                             1
paternity / maternity leave              1
retiree                               3443
student                                  1
unemployed                               2

There is no problem with the `income_type` column

### 2.3. Checking Duplicates

In [ ]:
# Checking for duplicates
print(df.duplicated().sum())
print()
df_duplicated_percent = df.duplicated().sum() / df.shape[0] 
print(f'Distribution of duplicate values by: {df_duplicated_percent:.2%}')

72

Distribution of duplicate values by: 0.33%


There are duplicate values of `72` or only `0.3%` in the dataset, so we decide to 'drop` for duplicate values because these values are still acceptable.

In [ ]:
# Address duplicates, if any
df = df.drop_duplicates().reset_index(drop=True)

In [ ]:
# Finally check if we have duplicates
df.duplicated().sum()

0

In [ ]:
# Check the size of the dataset you now have after the first manipulation you performed
df.shape

(21452, 12)

In [ ]:
# return the ratio of old and new data
old_df_shape = 21525
df_shape_percentage = (old_df_shape - df.shape[0]) / old_df_shape
print(f'The percentage of change in the dataset is: {df_shape_percentage:.2%}')

The percentage of change in the dataset is: 0.34%


We've fixed some issues in the dataset, such as:
1. Negative and overestimated values in the `days_employed` column.
2. Irregular register in the `education` column.
3. Remove duplicate values.
4. Some problems occurred in the `children` and `gender` columns.

So we get a change of `0.34%`, which means that this value has no significant impact on the dataset because it is less than `1%`.

### 2.4. Work with Missing Values

We include the `numpy` dictionary to speed up our work which will be used to `replace` the value `0` in the `days_employed` column after creating several age categories.

In [ ]:
# Import dictionary
import numpy as np

#### 2.4.1. Fixed Missing Values in `total_income`

We will start to work out the missing values in the `total_income` column.

We will first create several categories based on the age range from the `dob_years` column, which hopefully will help us find the average value of each age range which we will use to fill in the missing values for each category.

In [ ]:
# Write a function to calculate age categories
def age_group(age):
 
    if age <= 30:
        return '19-30'
    elif age <= 40:
        return '31-40'
    elif age <= 50:
        return '41-50'
    elif age <= 60:
        return '51-60'
    else:
        return '+60'

In [ ]:
# Test whether the function works or not
print(age_group(23))
print(age_group(38))
print(age_group(46))
print(age_group(55))
print(age_group(70))

19-30
31-40
41-50
51-60
+60


In [ ]:
# Create new column based on function
df['age_group'] = df['dob_years'].apply(age_group)

In [ ]:
# Memeriksa bagaimana nilai di dalam kolom baru
df.tail(10)

children  days_employed  dob_years            education  education_id  \
21442         1     467.685130         28  secondary education             1   
21443         0     914.391429         42    bachelor's degree             0   
21444         0     404.679034         42    bachelor's degree             0   
21445         0       0.000000         59  secondary education             1   
21446         1    2351.431934         37      graduate degree             4   
21447         1    4529.316663         43  secondary education             1   
21448         0       0.000000         67  secondary education             1   
21449         1    2113.346888         38  secondary education             1   
21450         3    3112.481705         38  secondary education             1   
21451         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21442            married                 0      F    employee     1   
21443            married                 0      F    business     0   
21444  civil partnership                 1      F    business     0   
21445            married                 0      F     retiree     0   
21446           divorced                 3      M    employee     0   
21447  civil partnership                 1      F    business     0   
21448            married                 0      F     retiree     0   
21449  civil partnership                 1      M    employee     1   
21450            married                 0      M    employee     1   
21451            married                 0      F    employee     0   

       total_income                     purpose age_group  
21442     17517.812          to become educated     19-30  
21443     51649.244    purchase of my own house     41-50  
21444     28489.529           buying my own car     41-50  
21445     24618.344           purchase of a car     51-60  
21446     18551.846  buy commercial real estate     31-40  
21447     35966.698        housing transactions     41-50  
21448     24959.969           purchase of a car       +60  
21449     14347.610                    property     31-40  
21450     39054.888           buying my own car     31-40  
21451     13127.587                to buy a car     31-40

- Some factors that can affect income include `education`, `days_employed` (work experience in days), and `income_type` (type of work).
- To determine whether we will use the `mean` or `median` to fill in the missing values, we conduct a deeper exploration to find out how the distribution of the `mean` and `median` values based on the categories we group is normally distributed or not.

In [ ]:
# Create a table without missing values and display some of its rows
df_clean = df[df.notna()]
df_clean.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0       0.000000         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose age_group  
0                purchase of the house     41-50  
1                         car purchase     31-40  
2                purchase of the house     31-40  
3              supplementary education     31-40  
4                    to have a wedding     51-60  
5                purchase of the house     19-30  
6                 housing transactions     41-50  
7                            education     41-50  
8                     having a wedding     31-40  
9  purchase of the house for my family     41-50

In [ ]:
# Display the 'mean' value of 'total_income' based on 'age_group'
df_clean.groupby(['age_group'])['total_income'].mean()

age_group
+60      23057.777452
19-30    25815.651899
31-40    28376.735148
41-50    28332.806009
51-60    25482.856294
Name: total_income, dtype: float64

In [ ]:
# Display the 'median' value of 'total_income' by 'age_group'
df_clean.groupby(['age_group'])['total_income'].median()

age_group
+60      19637.0560
19-30    22955.4740
31-40    24825.1865
41-50    24563.6500
51-60    22056.7710
Name: total_income, dtype: float64

The `mean` category-based average income has a larger average for each category, so I will choose the `median` because it better represents the category average by age group.

In [ ]:
# It's time we fill in the missing values
df['total_income'] = df.groupby(['age_group'])['total_income'].transform(lambda x: x.fillna(x.median()))

In [ ]:
# Check descriptive statistics of 'total_income'
df['total_income'].describe()

count     21452.000000
mean      26447.583422
std       15690.140709
min        3306.762000
25%       17217.441750
50%       23234.038000
75%       31328.693750
max      362496.645000
Name: total_income, dtype: float64

In [ ]:
# Check if missing values are filled
df['total_income'].isna().sum()

0

It is known that the missing data has not been filled in the `total_income` column because we have not entered the value from the new column that we created to replace the missing value from that column. To fill in the column, we will replace it with the `fillna` method of the new column we created, namely `total_revenue`.

As we can see, there are no missing values from the `total_income` column.
Time to check if the sum of the `total_income` columns is the same as the other columns.

In [ ]:
# Checking the number of column entries from the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21452 entries, 0 to 21451
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21452 non-null  int64  
 1   days_employed     19350 non-null  float64
 2   dob_years         21452 non-null  int64  
 3   education         21452 non-null  object 
 4   education_id      21452 non-null  int64  
 5   family_status     21452 non-null  object 
 6   family_status_id  21452 non-null  int64  
 7   gender            21452 non-null  object 
 8   income_type       21452 non-null  object 
 9   debt              21452 non-null  int64  
 10  total_income      21452 non-null  float64
 11  purpose           21452 non-null  object 
 12  age_group         21452 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


- From the information above, the amount in the `total_income` column already has the same value as the other columns.
- Next, we will fix the value in the `days_employed` column.

#### 2.4.2. Fixed the Missing Values in `days_employed`

Of course, the most influential factor in the `days_employed` column is the `age_category`. Imagine that it is impossible for someone to work beyond the legal age of a person or even to have work experience beyond their own age; that is impossible.

In [ ]:
# Display the 'median' value of 'days_employed' by 'age_group'
df_clean.groupby(['age_group'])['days_employed'].median()

age_group
+60         0.000000
19-30    1041.692949
31-40    1601.812856
41-50    1931.819208
51-60     551.486703
Name: days_employed, dtype: float64

In [ ]:
# Display the 'median' value of 'days_employed' by 'age_group'
df_clean.groupby(['age_group'])['days_employed'].mean()

age_group
+60       799.550825
19-30    1276.700146
31-40    2075.768902
41-50    2685.878763
51-60    1937.100928
Name: days_employed, dtype: float64

We think the `mean` could represent the value for the `days_employed` column, except that `median` has a value of `0`, which may be because we changed the value to be too high in the previous test, but does someone who has reached retirement age have no work experience at all? Or maybe someone with retired experience would count as a `0` value.

In [ ]:
# Time for us to replace the missing values and the '0' value with the average
df.loc[df['days_employed'] == 0, 'days_employed'] = np.NaN
df['days_employed'] = df.groupby(['age_group'])['days_employed'].transform(lambda x: x.fillna(x.mean()))

In [ ]:
# Display descriptive statistics of the column 'days_employed'
df['days_employed'].describe()

count    21452.000000
mean      2560.787849
std       2049.095043
min         24.141633
25%       1023.688788
50%       2228.980549
75%       3332.487952
max      18388.949901
Name: days_employed, dtype: float64

In [ ]:
# Memeriksa apakah nilai yang hilang telah teratasi
df['days_employed'].isna().sum()

0

Checking the number of columns of the entire dataset.

In [ ]:
# Checking the information of the entire dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21452 entries, 0 to 21451
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21452 non-null  int64  
 1   days_employed     21452 non-null  float64
 2   dob_years         21452 non-null  int64  
 3   education         21452 non-null  object 
 4   education_id      21452 non-null  int64  
 5   family_status     21452 non-null  object 
 6   family_status_id  21452 non-null  int64  
 7   gender            21452 non-null  object 
 8   income_type       21452 non-null  object 
 9   debt              21452 non-null  int64  
 10  total_income      21452 non-null  float64
 11  purpose           21452 non-null  object 
 12  age_group         21452 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


All columns, including `days_employed`, have the same value. We value that our work for *data cleansing* has been completed. Let's explore the data further to get interesting things in the data and our reference in making decisions.

### 2.5. Data Categorization

It seems that we found something interesting in the `purpose` column. That is, we value many data categories, and we can simplify them to be more general so that it will be easier for us to carry out investigations in making decisions.

#### 2.5.1. Categorizing Data in the `purpose` Column

In [ ]:
# Display a 'purpose' column to categorize with more generality
df['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            767
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             620
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

Checking the `unique` value.

In [ ]:
# Checking for unique values
df['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

From our previous observations, the categorization that occurs in the `purpose` column can be generalized into categories that are general and easy to understand, such as `car`, `property`, `education`, and `wedding`. Let's create the function below:

In [ ]:
# Write a function to categorize data based on common topics
def purpose_common(purpose):
    
    if 'property' in purpose:
        return 'property'
    elif 'estate' in purpose:
        return 'property'
    elif 'hous' in purpose:
        return 'property'
    elif 'car' in purpose:
        return 'car'
    elif 'educ' in purpose:
        return 'education'
    elif 'univ' in purpose:
        return 'education'
    elif 'wedd' in purpose:
        return 'wedding'

In [ ]:
# Load columns with categories and calculate their values
df['general_purpose'] = df['purpose'].apply(purpose_common)
print(df['general_purpose'].value_counts())
print()
df['general_purpose'].count()

property     10810
car           4306
education     4013
wedding       2323
Name: general_purpose, dtype: int64



21452

#### 2.5.2. Categorizing Data in the `total_income` Column

We will create categories based on `total_income` into several classes.

In [ ]:
# View the 'total_income' column to be categorized
df['total_income'].value_counts().sort_index()

3306.762      1
3392.845      1
3418.824      1
3471.216      1
3503.298      1
             ..
273809.483    1
274402.943    1
276204.162    1
352136.354    1
362496.645    1
Name: total_income, Length: 19348, dtype: int64

In [ ]:
# Get statistical inference for the 'total_income' column
df['total_income'].describe()

count     21452.000000
mean      26447.583422
std       15690.140709
min        3306.762000
25%       17217.441750
50%       23234.038000
75%       31328.693750
max      362496.645000
Name: total_income, dtype: float64

We categorize `total_income` into several categories based on economic level classification. This will facilitate us in the decision-making process in the future. Let's create the function below:

In [ ]:
# Create a function for categorizing into class groups
def income_class(total_income):
 
    if total_income <= 32000:
        return 'poor'
    elif total_income <= 53000:
        return 'lower middle class'
    elif total_income <= 106000:
        return 'middle class'
    elif total_income <= 373000:
        return 'upper middle class'
    else:
        return 'rich'

In [ ]:
# Create a new column with categories
df['economic_class'] = df['total_income'].apply(income_class)

In [ ]:
# Calculating the distribution
df['economic_class'].value_counts()

poor                  16387
lower middle class     3985
middle class           1000
upper middle class       80
Name: economic_class, dtype: int64

## 3. Answering the Hypothesis

### 3.1. Is there a relationship between having children and the probability of defaulting on loans?

In [ ]:
# Checking if the 'children' column affects 'debt'
pivot_table_children = df.pivot_table(index='children', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_children

debt            0       1
children                 
0         13026.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

In [ ]:
# Checking percentages to get conclusions
pivot_table_children['percent_1'] = pivot_table_children[1] / (pivot_table_children[1] + pivot_table_children[0]) * 100
pivot_table_children

debt            0       1  percent_1
children                            
0         13026.0  1063.0   7.544893
1          4410.0   445.0   9.165808
2          1926.0   202.0   9.492481
3           303.0    27.0   8.181818
4            37.0     4.0   9.756098
5             9.0     NaN        NaN

**Conclusion**

From the data above, we find that:
1. Clients who have `1` to `4` children have almost the same percentage at `8%` to `9%`.
2. Clients who have `5` children do not have loan payment debts, but we cannot use them as a reference because the data is too small.
3. Clients who **don't** have children have the lowest ratio for loan repayments of `7%`. This may be because they have fewer dependents than clients with children.

This will certainly make it easier for us to make decisions in providing credit to clients who do not have children because of their ability to pay off their credit.

### 3.2. Is there a correlation between family status and paying back on time?

In [ ]:
# Checking the 'family_status' data affects the 'debt' column
pivot_table_family_status = df.pivot_table(index='family_status', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_family_status

debt                   0    1
family_status                
civil partnership   3761  388
divorced            1110   85
married            11408  931
unmarried           2536  274
widow / widower      896   63

In [ ]:
# Calculate the ratio to find the conclusion
pivot_table_family_status['percent_1'] = pivot_table_family_status[1] / (pivot_table_family_status[1] + pivot_table_family_status[0]) * 100
pivot_table_family_status

debt                   0    1  percent_1
family_status                           
civil partnership   3761  388   9.351651
divorced            1110   85   7.112971
married            11408  931   7.545182
unmarried           2536  274   9.750890
widow / widower      896   63   6.569343

**Conclusion**

There are some interesting things that we find here:
1. Clients who are `unmarried` and in `civil partnership` have a fairly high percentage of `9%`.
2. Clients who are `divorced` and `married` have a ratio of `7%`, which means it is smaller than clients who are not married. Is it because they can combine income with their partners? This is inversely proportional to the previous test that clients who do not have children have a smaller percentage of debt, considering that someone married has a tendency to have children.
3. Clients with `widow/widower` status have the smallest debt percentage of `6%`. Are we going to consider status as a basis for decision-making?

### 3.3. Is there a correlation between income levels and paying back on time?

In [ ]:
# Check if 'economic_class' has a correlation with 'debt'
pivot_table_economic_class = df.pivot_table(index='economic_class', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_economic_class

debt                    0     1
economic_class                 
lower middle class   3704   281
middle class          928    72
poor                15004  1383
upper middle class     75     5

In [ ]:
# Calculate distribution percentages to draw conclusions
pivot_table_economic_class['percent_1'] = pivot_table_economic_class[1] / (pivot_table_economic_class[1] + pivot_table_economic_class[0]) * 100
pivot_table_economic_class

debt                    0     1  percent_1
economic_class                            
lower middle class   3704   281   7.051443
middle class          928    72   7.200000
poor                15004  1383   8.439617
upper middle class     75     5   6.250000

**Conclusion**

Some of the things we find from the manipulation of the data above include the following:
1. Clients with `lower middle class` and `middle class` economic levels have a balanced percentage of `7%` for clients with credit payment debts.
2. Clients with a `poor` economic level can be in greater arrears, namely `8%`. Can this affect our decision not to give credit, given that there are so many of them?
3. Clients with `upper middle class` income have the least risk of `6%`.

### 3.4. How do different loan objectives affect the probability of defaulting on a loan??

In [ ]:
# Checking the percentage of 'general_purpose' to 'debt' to explore conclusions
pivot_table_general_purpose = df.pivot_table(index='general_purpose', columns= 'debt', values='days_employed', aggfunc='count')
pivot_table_general_purpose['percent_1'] = pivot_table_general_purpose[1] / (pivot_table_general_purpose[1] + pivot_table_general_purpose[0]) * 100
pivot_table_general_purpose

debt                 0    1  percent_1
general_purpose                       
car               3903  403   9.359034
education         3643  370   9.220035
property         10028  782   7.234043
wedding           2137  186   8.006888

**Conclusion**

Based on the handling, we do the things we can get:
1. Clients who make loans for `car` and `education` purposes have the highest default risk of `9%`.
2. Clients who use their loans for `wedding` have a lower percentage of loan debt than the previous criteria.
3. Clients who pay on time, namely for the use of `property`, with a risk of non-current debt of `7%`.

## 4. General Conclusion

- We have done a *cleansing data* process to fix problematic data in our dataset. The cleaning that we do includes filling in missing values, removing duplicate values, correcting irregular registers and values that are too large, and replacing unreasonable values so that we get a dataset we can process for the credit analysis process.

- The findings that we get after exploring are that there is a correlation between the number of children and marital status in the risk of credit payments. Clients who do not have children will find it easier to pay off their debts than clients with children. Clients who are married or have had a partner have a lower risk of default than clients who are single or live together.
Clients with lower incomes will be more likely to have loan debt, and clients who use the money for home purposes will have a higher percentage of them being able to pay off their debts.

- But can we use all the data manipulation we do in the **decision-making** process to minimize the risks that will occur in the future?